Managing vacation/holidays in bulk

apooravc opened the pull request:
This first commit adds the interface for this feature where user can enter holidays, vacations in bulk without going to the dates on the calendar to do it. Calendar now has an _add_ button as shown below:
![add](https://user-images.githubusercontent.com/25399108/41508393-aa5ab190-7261-11e8-9f34-60ba5bcee633.JPG)

Clicking on this button would open a dialog modal having two tab views/forms:

Add Vacation form
![vac](https://user-images.githubusercontent.com/25399108/41508397-cea56c2a-7261-11e8-842a-20585c2089f0.JPG)

Functionality:
In this tab, user can select a range of consecutive dates of vacation for a selected provider from _Start Date_ to _End Date_. Upon submitting this form, vacation slot would be booked for specified dates and specified provider.


Add Holiday form
![holid](https://user-images.githubusercontent.com/25399108/41508399-d95d5cc2-7261-11e8-934c-32082969c581.JPG)

Functionality:
In this tab, user can select and _add_ disjoint dates of holiday (which will be shown under _Selected dates_ column) for a particular clinic/facility. Upon submitting this form, holiday slot would be booked for specified dates and all providers related with that clinic/facility.

@teryhill Please have a look.

teryhill commented:
I am guessing that this is just a UI since it did not put anything on the calendar. 

The Holiday will apply to the entire clinic , it should also allow 1/2 days. There needs to be a check on both functions to make sure no appointment is scheduled for any of the requested dates.

aethelwulffe commented:
Can you do the sanity check for existing appointments that have not been canceled/moved?  Should be good otherwise.

apooravc commented:
@aethelwulffe You're referring to the check which Terry mentioned right? Check to make sure no appointment is scheduled for any of the requested dates,, before forming Vacation/Holiday slot on them?

aethelwulffe commented:
Yes.  That thing.  Should be at least a warning.  I can totally see this scrumming up a well meaning admin, while causing the front desk folks to want to kill them....and some pretty upset patients.  At very least, it is a nice additional tool for ways to FIND blocks of time to schedule the holiday.  We have checks like that already I believe.

teryhill commented:
@apooravc  it put the vacation on the scheduled still not closing the popup, you are getting there.

apooravc commented:
Last three commits added back-end code to create event slots according to Provider Vacation dates, radio buttons to Clinic holidays for allowing half days.

Clinic Holiday preview:
![holiday-form](https://user-images.githubusercontent.com/25399108/42005709-8c8a4272-7a93-11e8-98d8-d7fe45d65b2e.JPG)

Creating events using Provider Vacation:
![vacation](https://user-images.githubusercontent.com/25399108/42005730-a9bd1108-7a93-11e8-90b5-a7be561ba3e2.JPG)

Yet to address/fix:
1. Clinic holiday code. (Trying to pass dates and day types)
2. Closing of Add Vacation/Holiday window after submission.

@teryhill Please have a look.

apooravc commented on modules/calendar/add.php:
```diff
@@ -0,0 +1,354 @@
+                                                 WHERE u.id = ?", array($_SESSION['authId']) ));
+    // default_user_facility denotes id of default facility of logged in user
+    $default_user_facility = $def_facility['facility_id'];
+    // Taking facility & billing facility as default facility of logged in user
```
Since provider is on a vacation, I assumed this. Is this assumption fine?

teryhill replied:
Yes that is fine . They should only be working with the facility that they have access to.

apooravc commented:
@teryhill Added code for Clinic Holidays case. It takes a date (key) & day (value) type (full day/half day) and then for each provider, creates a holiday event.

teryhill commented:
Tested and it works Great @apooravc . Can you go a step further and check the categories to see if they are a clinic option and put them in a drop down?

![image](https://user-images.githubusercontent.com/4655366/42063525-1d36c1b2-7b00-11e8-9000-c58500df48be.png)

teryhill commented:
Yes but make it a drop down. If the user wants another category say "Blue Grass Day" they can and close the clinic for the event.

teryhill commented:
@apooravc hold on and let me think that thru,

teryhill commented:
@aethelwulffe @tmccormi I am thinking this thru and really instead of managing Holidays we are actually managing clinic closure. So instead on putting it on the schedule as holidays Should we use "Closed" instead and instead of "Add" on the calendar what about "provider/facility". With the intent to move all the provider options from where they are currently to here. Then ACL can be applied easier.

Thoughts?

teryhill commented:
What about Lunch? I think you are allowing the "Lunch" to produce an error.

apooravc commented:
@teryhill I tested sanity check with event category as "Lunch", it was working. Could you explain that user case and error? I noticed that I should also add provider id in case of checking adding vacation in `fetchAllEvents($start_date, $end_date)` because it is possible that provider B has some events b/w vacation range dates for provider A and this check will not let user add vacations in such case.

teryhill commented:
The check for provider needs to be done in both instances. I an getting the alert when I try to add holidays because the in office and out of office tags. @apooravc

apooravc commented:
@teryhill Isn't a holiday independent of provider, like if I make a holiday event on a date, it will be for all providers and it should check if no provider has any event scheduled which conflicts with holiday event on that day? So, what type of case would require inclusion of a provider check in Add Holiday?

teryhill commented:
@apooravc you are currently checking to see if any events are scheduled on the holiday date. The issue is Lunch , In office and out of office show up as events so you cannot schedule a holiday . The check needs to be is there any patient appointments on the day. PID != 0

apooravc commented:
Latest commit adds provider checks in both Add Vacation and Add Holidays options. With this:
1. In case of former, it restricts check to selected provider for which vacation is added.
2. In case of latter, it checks if on at least one of selected holiday date, at least one event is patient event then it will not add holidays for selected dates. If all events are provider events, it will add holidays.

@teryhill Please have a look,

apooravc commented:
Latest commit adds:
1. Ability to make holiday events by adding a specific facility.
2. Ability to delete a particular set of holidays by adding corresponding date and facility.

@teryhill Please have a look.

aethelwulffe commented:
I am going to apply this without and on top of some other PR commits to test.  Scheduled for 16:00 today.

teryhill commented on modules/calendar/add.php:
```diff
@@ -0,0 +1,548 @@
+    foreach ($dateAndType as $date => $value) {
+      // for each set of date and facility, delete holiday event under all providers
+      // no need to iterate through providers in this case
+      sqlStatement("DELETE FROM libreehr_postcalendar_events WHERE pc_eventDate = ? AND pc_facility = ?", array($date, $value[1]));
```
You need to add the pc_catid here or else you might delete appointments that are there.

teryhill commented on modules/calendar/add.php:
```diff
@@ -0,0 +1,548 @@
+      if (!empty($fetchedEvents)) {
+        foreach ($fetchedEvents as $event) {
+          // check if event is a patient event
+          if (event['pc_pid'] > 0) {
```
Small bug is if you are adding for a specific facility and there is an appointment in another facility you get the message. Facility checking needs to be done.

apooravc replied:
@teryhill If a user adds a holiday for a particular date, and if an appointment is on that date even if in another facility, shouldn't it be prevented from adding so as to not overlap it?

teryhill replied:
If one facility needs to close it should not be prevented by another facilities appointments. They are separate. You can wait until customers bring up issues if you want.

teryhill replied:
Set up 2 facilities, add the facility to a user as the default. Add an appointment to that user's calendar. On the second facility add a holiday for the same day the appointment is on.

apooravc replied:
@teryhill It did the alert. So, overlapping of events in case when date is same, provider is same but facility is different should be fine? And it should only alert when holiday's and appt.'s date, provider and facility are same?

apooravc replied:
I think checking with date and facility is enough because if facility is closed, there can't be that appt. in any case, right?

teryhill replied:
if ($event['pc_pid'] != 0) {
              if($event['pc_facility'] == $value[1]) {

teryhill commented on modules/calendar/add.php:
```diff
@@ -0,0 +1,548 @@
+
+    // reached here, means no conflicting dates, make requested events
+    foreach ($dateAndType as $date => $value) {
+      $providers = getProviderInfo();
```
There is no facility check so the Holiday is applied to every facility, even if you select a facility it is still applied to all.

apooravc replied:
@teryhill Facility id is stored in `$value` corresponding to `$date`. I tested with different facilities and holiday was made w.r.t to selected facilities. It wasn't applied to all, am I doing something wrong here?

teryhill replied:
What is w.r.t?

apooravc replied:
with respect to

teryhill replied:
I have a provider that is only at one facility and other providers at another facility. If I select the facility that has one provider and make a holiday it makes it for "All" providers no matter which facility is selected.

apooravc replied:
Oh, so you are talking about it being applied to all providers when we make a holiday event. That's the default behavior because facility is closed no matter which provider.

teryhill replied:
What if they use this to close a facility because the facility is being repaired?

apooravc replied:
To add under only those providers which are at that facility selected for holiday, I think we have to take help of getUsersFacility.

apooravc replied:
@teryhill Will change adding under all providers to under only those providers specific to that selected facility.

teryhill commented on modules/calendar/add.php:
```diff
@@ -0,0 +1,548 @@
+        $args['locationspec'] = $locationspec;
+        // Taking facility & billing facility as selected facility of holiday
+        $args['facility'] = $value[1];
+        $args['billing_facility'] = $value[1];
```
Billing facility in non provider events is important and could be different from the facility. It does not matter with provider events just want you to know about it .

teryhill commented on modules/calendar/add.php:
```diff
@@ -0,0 +1,548 @@
+        <label class="control-label col-sm-2" for="clinic_id">Clinic</label>
+        <div class="col-sm-10">
+          <select class="form-control" id="clinic_id" name="clinic_id">
+            <option value="0" data-text="Undefined">Choose Clinic/Facility</option>
```
An option for "All" facilities would be nice to have.

apooravc replied:
@teryhill So, if user makes a holiday with a date, day type and facility option "All", what should be stored in DB in facility id? I mean how should it work?

teryhill replied:
The facilities listed in the facility table. You could wait on this also and see if the customers want it.

apooravc commented on modules/calendar/index.php:
```diff
@@ -138,6 +138,7 @@
       var title_agenda2 = '<?php echo xlt('2 day'); ?>';
       var title_agenda = '<?php echo xlt('1 day'); ?>';
       var title_search = '<?php echo xlt('search'); ?>';
+      var title_add = '<?php echo xlt('add'); ?>';
```
Calling it "Provider" would be good as it is also the tab name in event panel.

aethelwulffe replied:
Either one is fine,  and a reference to NPI is not international.  Provider/Facility is most completely descriptive, but long.

apooravc commented:
@teryhill Did some changes:
1. Added a check b/w holiday's facility and event's facility on same date.
2. Holiday would now add under providers which are specific to that facility of holiday.
3. Added translation identifiers.
4. Added holiday category in holiday deletion.

Please have a look.

aethelwulffe commented:
REstesting now.

aethelwulffe commented:
Nevermind.  I got to kill my error reporting to test this.  Putting it a couple of spots back in the queue.

aethelwulffe commented:
@teryhill Please go over this to check if all your points are met with the appointment/calendar schedule of holidays.  Lots of work just sitting here.

aethelwulffe commented:
These are all approved.  Three months without second approval or protest.  Merging.  Everyone, this is now up for general testing in the main branch.  Let me know if there are issues.  I did my test on this LOOONG ago.
